File: threads/src/main/java/im/threads/broadcastReceivers/ProgressReceiver.java
package im.threads.broadcastReceivers;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.util.Log;

import java.lang.ref.WeakReference;

import im.threads.fragments.ChatFragment;
import im.threads.model.FileDescription;
import im.threads.services.DownloadService;

/**
 * В чате есть возможность скачать файл из сообщения.
 * Он скачивается через сервис.
 * Для приема сообщений из сервиса используется данный BroadcastReceiver
 */
public class ProgressReceiver extends BroadcastReceiver {
    private static final String TAG = "ProgressReceiver ";
    // Сообщения для Broadcast Receivers
    public static final String PROGRESS_BROADCAST = "im.threads.controllers.PROGRESS_BROADCAST";
    public static final String DOWNLOADED_SUCCESSFULLY_BROADCAST = "im.threads.controllers.DOWNLOADED_SUCCESSFULLY_BROADCAST";
    public static final String DOWNLOAD_ERROR_BROADCAST = "im.threads.controllers.DOWNLOAD_ERROR_BROADCAST";
    public static final String DEVICE_ID_IS_SET_BROADCAST = "im.threads.controllers.DEVICE_ID_IS_SET_BROADCAST";

    private WeakReference<ChatFragment> fragment;
    private WeakReference<DeviceIdChangedListener> deviceIdChangedListener;

    public ProgressReceiver(ChatFragment fragment, DeviceIdChangedListener deviceIdChangedListener) {
        this.fragment = new WeakReference<>(fragment);
        this.deviceIdChangedListener = new WeakReference<>(deviceIdChangedListener);
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        Log.i(TAG, "onReceive:");
        String action = intent.getAction();
        if (action == null) return;
        if (action.equals(PROGRESS_BROADCAST)) {
            Log.i(TAG, "onReceive: PROGRESS_BROADCAST ");
            FileDescription fileDescription = intent.getParcelableExtra(DownloadService.FD_TAG);
            if (fragment != null && fileDescription != null)
                fragment.get().updateProgress(fileDescription);
        } else if (action.equals(DOWNLOADED_SUCCESSFULLY_BROADCAST)) {
            Log.i(TAG, "onReceive: DOWNLOADED_SUCCESSFULLY_BROADCAST ");
            FileDescription fileDescription = intent.getParcelableExtra(DownloadService.FD_TAG);
            fileDescription.setDownloadProgress(100);
            if (fragment != null)
                fragment.get().updateProgress(fileDescription);
        } else if (action.equals(DOWNLOAD_ERROR_BROADCAST)) {
            Log.i(TAG, "onReceive: DOWNLOAD_ERROR_BROADCAST ");
            FileDescription fileDescription = intent.getParcelableExtra(DownloadService.FD_TAG);
            if (fragment != null && fileDescription != null) {
                Throwable t = (Throwable) intent.getSerializableExtra(DOWNLOAD_ERROR_BROADCAST);
                fragment.get().onDownloadError(fileDescription, t);
            }
        } else if (action.equals(DEVICE_ID_IS_SET_BROADCAST)) {
            deviceIdChangedListener.get().onSettingClientId(context);
        }
    }

    public interface DeviceIdChangedListener {
        void onSettingClientId(final Context ctx);
    }
}

File: src/main/java/im/threads/business/broadcastReceivers/ProgressReceiver.kt
package im.threads.business.broadcastReceivers

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import im.threads.business.logger.LoggerEdna
import im.threads.business.models.FileDescription
import im.threads.business.workers.FileDownloadWorker
import java.lang.ref.SoftReference

/**
 * В чате есть возможность скачать файл из сообщения.
 * Он скачивается через сервис.
 * Для приема сообщений из сервиса используется данный BroadcastReceiver
 */
class ProgressReceiver(callback: Callback) : BroadcastReceiver() {
    private val callback = SoftReference(callback)

    override fun onReceive(context: Context, intent: Intent) {
        val action = intent.action ?: return
        when (action) {
            PROGRESS_BROADCAST -> {
                intent.getParcelableExtra<FileDescription>(FileDownloadWorker.FD_TAG)?.let {
                    callback.get()?.updateProgress(it)
                }
            }
            DOWNLOADED_SUCCESSFULLY_BROADCAST -> {
                LoggerEdna.info("onReceive: DOWNLOADED_SUCCESSFULLY_BROADCAST ")
                intent.getParcelableExtra<FileDescription>(FileDownloadWorker.FD_TAG)?.let {
                    it.downloadProgress = 100
                    callback.get()?.updateProgress(it)
                }
            }
            DOWNLOAD_ERROR_BROADCAST -> {
                LoggerEdna.error("onReceive: DOWNLOAD_ERROR_BROADCAST ")
                intent.getParcelableExtra<FileDescription>(FileDownloadWorker.FD_TAG)?.let {
                    callback.get()?.onDownloadError(
                        it,
                        intent.getSerializableExtra(DOWNLOAD_ERROR_BROADCAST) as Throwable?
                    )
                }
            }
        }
    }

    interface Callback {
        /**
         * Определяет реакцию на обновление прогресса
         * @param fileDescription характеристики файла, где произошел прогресс загрузки
         */
        fun updateProgress(fileDescription: FileDescription?)

        /**
         * Определяет реакцию на ошибку при загрузке
         * @param fileDescription характеристики файла, где произошла ошибка загрузки
         */
        fun onDownloadError(fileDescription: FileDescription?, throwable: Throwable?)
    }

    companion object {
        // Сообщения для Broadcast Receivers
        const val PROGRESS_BROADCAST = "im.threads.internal.controllers.PROGRESS_BROADCAST"
        const val DOWNLOADED_SUCCESSFULLY_BROADCAST =
            "im.threads.internal.controllers.DOWNLOADED_SUCCESSFULLY_BROADCAST"
        const val DOWNLOAD_ERROR_BROADCAST =
            "im.threads.internal.controllers.DOWNLOAD_ERROR_BROADCAST"
    }
}
